core/internal/filetransfer/file_transfer_default_test.go

```diff
@@ -105,3 +107,82 @@ func TestDefaultFileTransfer_Upload(t *testing.T) {
 	err = ft.Upload(task)
 	assert.NoError(t, err)
 }
+
+func TestDefaultFileTransfer_UploadNotFound(t *testing.T) {
+	fnfHandler := func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusNotFound)
+	}
+	err := uploadToServerWithHandler(t, fnfHandler)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "404")
+}
+
+func TestDefaultFileTransfer_UploadConnectionClosed(t *testing.T) {
+	closeHandler := func(w http.ResponseWriter, r *http.Request) {
+		hj, ok := w.(http.Hijacker)
+		assert.True(t, ok, "webserver doesn't support hijacking")
+		conn, _, err := hj.Hijack()
+		assert.NoError(t, err, "hijacking error")
+		conn.Close()
+	}
+	err := uploadToServerWithHandler(t, closeHandler)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "giving up after 2 attempt(s)")
+}
+
+func TestDefaultFileTransfer_UploadNoResponse(t *testing.T) {
+	err := uploadToServerWithHandler(t, func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(500 * time.Millisecond)
+	})
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "context deadline exceeded")
+}
+
+func TestDefaultFileTransfer_UploadNoServer(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
+	ft := filetransfer.NewDefaultFileTransfer(observability.NewNoOpLogger(), impatientClient())
+
+	tempFile, err := os.CreateTemp("", "")
+	assert.NoError(t, err)
+	defer os.Remove(tempFile.Name())
+
+	task := &filetransfer.Task{
+		Type: filetransfer.UploadTask,
+		Path: tempFile.Name(),
+		Url:  server.URL,
+	}
+
+	// Close the server before the upload begins.
+	server.Close()
+
+	err = ft.Upload(task)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "connection refused")
+}
+
+func uploadToServerWithHandler(t *testing.T, handler func(w http.ResponseWriter, r *http.Request)) error {
```

Comment:
nit: line length

```suggestion
func uploadToServerWithHandler(
	t *testing.T,
	handler func(w http.ResponseWriter, r *http.Request),
) error {
```

Reply:
Does `gofmt` not format code the way we want?

Reply:
`gofmt` is different from formatters for other languages in that it respects the author's line breaks. It gives more control to the author, but the downside is that some manual formatting is necessary